From e25c418345f0a28fa31f36f725e5f7c696e514b8 Mon Sep 17 00:00:00 2001
From: Ananya Ghosh <ananghos@qti.qualcomm.com>
Date: Wed, 22 Apr 2026 14:42:44 +0530
Subject: [PATCH] soc: qcom: ice: Introduce ops-based backend framework for SCM
 and OP-TEE

Refactor the Qualcomm Inline Crypto Engine (ICE) driver to support
multiple backend implementations through an ops-based architecture.
This enables runtime selection between TrustZone/SCM and OP-TEE
backends without compile-time dependencies or device tree configuration.

Change-Id: I8216284a6cfc8286994befa3b7ba45af75b8f388
Signed-off-by: Md Sadre Alam <mdalam@qti.qualcomm.com>
Signed-off-by: Ananya Ghosh <ananghos@qti.qualcomm.com>
---
 drivers/firmware/Kconfig               |  27 ++++
 drivers/firmware/Makefile              |   3 +
 drivers/firmware/qcom_scm.c            |  63 +--------
 drivers/md/dm-crypt.c                  |   4 +-
 drivers/soc/qcom/ice.c                 | 183 +------------------------
 include/linux/firmware/qcom/qcom_scm.h |   9 ++
 include/soc/qcom/ice.h                 |  10 +-
 7 files changed, 58 insertions(+), 241 deletions(-)

diff --git a/drivers/firmware/Kconfig b/drivers/firmware/Kconfig
index 3b18c33c7bdd..8091a48e9621 100644
--- a/drivers/firmware/Kconfig
+++ b/drivers/firmware/Kconfig
@@ -228,6 +228,33 @@ config QCOM_PAS_TEE
 	  Enable the generic Peripheral Authentication Service (PAS) provided
 	  by the firmware TEE implementation as the backend.
 
+config QCOM_ICE_CORE
+	tristate "Qualcomm Inline Crypto Engine Core"
+	help
+	  Enable the generic Inline Crypto Engine (ICE) core layer.
+	  This acts as the common layer with different backends plugged
+	  in whether it's an SCM implementation or a TEE-based implementation.
+
+config QCOM_ICE_SCM
+	tristate "Qualcomm ICE SCM Backend"
+	depends on QCOM_SCM
+	select QCOM_ICE_CORE
+	default y if QCOM_SCM
+	help
+	  Enable the SCM backend for Qualcomm Inline Crypto Engine.
+	  This backend uses Secure Monitor Calls (SMC) to communicate
+	  with TrustZone for key management operations.
+
+config QCOM_ICE_TEE
+	tristate "Qualcomm ICE TEE Backend"
+	depends on TEE
+	select QCOM_ICE_CORE
+	default m
+	help
+	  Enable the TEE backend for Qualcomm Inline Crypto Engine.
+	  This backend uses the TEE bus framework to communicate with
+	  OPTEE for key management operations.
+
 config QCOM_SCM
 	select QCOM_PAS
 	tristate
diff --git a/drivers/firmware/Makefile b/drivers/firmware/Makefile
index 471c16a24aa0..fe56c2233140 100644
--- a/drivers/firmware/Makefile
+++ b/drivers/firmware/Makefile
@@ -22,6 +22,9 @@ obj-$(CONFIG_QCOM_SCM)		+= qcom-scm.o
 qcom-scm-objs += qcom_scm.o qcom_scm-smc.o qcom_scm-legacy.o
 obj-$(CONFIG_QCOM_PAS)		+= qcom_pas.o
 obj-$(CONFIG_QCOM_PAS_TEE)	+= qcom_pas_tee.o
+obj-$(CONFIG_QCOM_ICE_CORE)	+= qcom_ice_core.o
+obj-$(CONFIG_QCOM_ICE_SCM)	+= qcom_ice_scm.o
+obj-$(CONFIG_QCOM_ICE_TEE)	+= qcom_ice_tee.o
 obj-$(CONFIG_SYSFB)		+= sysfb.o
 obj-$(CONFIG_SYSFB_SIMPLEFB)	+= sysfb_simplefb.o
 obj-$(CONFIG_TI_SCI_PROTOCOL)	+= ti_sci.o
diff --git a/drivers/firmware/qcom_scm.c b/drivers/firmware/qcom_scm.c
index 8eedf6dfe982..7e334e26ab06 100644
--- a/drivers/firmware/qcom_scm.c
+++ b/drivers/firmware/qcom_scm.c
@@ -327,7 +327,7 @@ static bool __qcom_scm_is_call_available(struct device *dev, u32 svc_id,
 	return ret ? false : !!res.result[0];
 }
 
-int qcom_context_ice_sec(u32 type, u8 key_size,
+int __qcom_context_ice_sec(u32 type, u8 key_size,
 			 u8 algo_mode, u8 *data_ctxt, u32 data_ctxt_len,
 			 u8 *salt_ctxt, u32 salt_ctxt_len)
 {
@@ -393,9 +393,8 @@ int qcom_context_ice_sec(u32 type, u8 key_size,
 	return ret ?  : res.result[0];
 
 }
-EXPORT_SYMBOL_GPL(qcom_context_ice_sec);
 
-int qcom_config_sec_ice(void *buf, int size)
+int __qcom_config_sec_ice(void *buf, int size)
 {
 	int ret;
 	dma_addr_t conf_phys;
@@ -415,7 +414,6 @@ int qcom_config_sec_ice(void *buf, int size)
 	ret = qcom_scm_call(__scm->dev, &desc, &res);
 	return ret ? false : !!res.result[0];
 }
-EXPORT_SYMBOL_GPL(qcom_config_sec_ice);
 
 static int qcom_scm_set_boot_addr(void *entry, const u8 *cpu_bits)
 {
@@ -1458,33 +1456,19 @@ int qcom_scm_ocmem_unlock(enum qcom_scm_ocmem_client id, u32 offset, u32 size)
 }
 EXPORT_SYMBOL_GPL(qcom_scm_ocmem_unlock);
 
-/**
- * qcom_scm_ice_available() - Is the ICE key programming interface available?
- *
- * Return: true iff the SCM calls wrapped by qcom_scm_ice_invalidate_key() and
- *	   qcom_scm_ice_set_key() are available.
- */
-bool qcom_scm_ice_available(void)
+bool __qcom_scm_ice_available(void)
 {
 	return __qcom_scm_is_call_available(__scm->dev, QCOM_SCM_SVC_ES,
 					    QCOM_SCM_ES_INVALIDATE_ICE_KEY) &&
 		__qcom_scm_is_call_available(__scm->dev, QCOM_SCM_SVC_ES,
 					     QCOM_SCM_ES_CONFIG_SET_ICE_KEY);
 }
-EXPORT_SYMBOL_GPL(qcom_scm_ice_available);
 
-/**
- * qcom_scm_ice_hwkey_available() - Is the ICE HW key programming
- *                                  interface available?
- *
- * Return: true if the SCM calls wrapped by qcom_config_sec_ice() are available.
- */
-bool qcom_scm_ice_hwkey_available(void)
+bool __qcom_scm_ice_hwkey_available(void)
 {
 	return __qcom_scm_is_call_available(__scm->dev, QCOM_SVC_ICE,
 					    QCOM_SCM_ICE_CMD);
 }
-EXPORT_SYMBOL(qcom_scm_ice_hwkey_available);
 
 /**
  * qcom_qfprom_show_auth_available() - Check if the SCM call to verify
@@ -1545,19 +1529,7 @@ bool qcom_qfrom_fuse_row_read_available(void)
 }
 EXPORT_SYMBOL_GPL(qcom_qfrom_fuse_row_read_available);
 
-/**
- * qcom_scm_ice_invalidate_key() - Invalidate an inline encryption key
- * @index: the keyslot to invalidate
- *
- * The UFSHCI and eMMC standards define a standard way to do this, but it
- * doesn't work on these SoCs; only this SCM call does.
- *
- * It is assumed that the SoC has only one ICE instance being used, as this SCM
- * call doesn't specify which ICE instance the keyslot belongs to.
- *
- * Return: 0 on success; -errno on failure.
- */
-int qcom_scm_ice_invalidate_key(u32 index)
+int __qcom_scm_ice_invalidate_key(u32 index)
 {
 	struct qcom_scm_desc desc = {
 		.svc = QCOM_SCM_SVC_ES,
@@ -1569,30 +1541,8 @@ int qcom_scm_ice_invalidate_key(u32 index)
 
 	return qcom_scm_call(__scm->dev, &desc, NULL);
 }
-EXPORT_SYMBOL_GPL(qcom_scm_ice_invalidate_key);
 
-/**
- * qcom_scm_ice_set_key() - Set an inline encryption key
- * @index: the keyslot into which to set the key
- * @key: the key to program
- * @key_size: the size of the key in bytes
- * @cipher: the encryption algorithm the key is for
- * @data_unit_size: the encryption data unit size, i.e. the size of each
- *		    individual plaintext and ciphertext.  Given in 512-byte
- *		    units, e.g. 1 = 512 bytes, 8 = 4096 bytes, etc.
- *
- * Program a key into a keyslot of Qualcomm ICE (Inline Crypto Engine), where it
- * can then be used to encrypt/decrypt UFS or eMMC I/O requests inline.
- *
- * The UFSHCI and eMMC standards define a standard way to do this, but it
- * doesn't work on these SoCs; only this SCM call does.
- *
- * It is assumed that the SoC has only one ICE instance being used, as this SCM
- * call doesn't specify which ICE instance the keyslot belongs to.
- *
- * Return: 0 on success; -errno on failure.
- */
-int qcom_scm_ice_set_key(u32 index, const u8 *key, u32 key_size,
+int __qcom_scm_ice_set_key(u32 index, const u8 *key, u32 key_size,
 			 enum qcom_scm_ice_cipher cipher, u32 data_unit_size)
 {
 	struct qcom_scm_desc desc = {
@@ -1635,7 +1585,6 @@ int qcom_scm_ice_set_key(u32 index, const u8 *key, u32 key_size,
 	dma_free_coherent(__scm->dev, key_size, keybuf, key_phys);
 	return ret;
 }
-EXPORT_SYMBOL_GPL(qcom_scm_ice_set_key);
 
 /**
  * qcom_scm_hdcp_available() - Check if secure environment supports HDCP.
diff --git a/drivers/md/dm-crypt.c b/drivers/md/dm-crypt.c
index 06323ff5b245..cc55db5ace68 100644
--- a/drivers/md/dm-crypt.c
+++ b/drivers/md/dm-crypt.c
@@ -3305,7 +3305,7 @@ static int qcom_set_ice_context(struct dm_target *ti, u32 argc, char **argv)
 	if (argc > 8 && argv[8] != NULL && !strcmp(argv[8], "oemseed")) {
 		seedtype = OEM_SEED_TYPE;
 	} else {
-		ret = qcom_context_ice_sec(seedtype, key_size, algo_mode,
+		ret = qcom_ice_set_context(seedtype, key_size, algo_mode,
 				hex_data_context, hex_data_len,
 				hex_salt_context, hex_salt_len);
 		if (ret)
@@ -3352,7 +3352,7 @@ static int qcom_set_ice_context(struct dm_target *ti, u32 argc, char **argv)
 		buf = NULL;
 	}
 
-	ret = qcom_context_ice_sec(seedtype, key_size, algo_mode, hex_data_context,
+	ret = qcom_ice_set_context(seedtype, key_size, algo_mode, hex_data_context,
 				hex_data_len, hex_salt_context, hex_salt_len);
 	if (ret)
 		DMERR("%s: ice context configuration fail\n", __func__);
diff --git a/drivers/soc/qcom/ice.c b/drivers/soc/qcom/ice.c
index 91409201dafd..217f5cdef440 100644
--- a/drivers/soc/qcom/ice.c
+++ b/drivers/soc/qcom/ice.c
@@ -15,16 +15,9 @@
 #include <linux/of.h>
 #include <linux/of_platform.h>
 #include <linux/platform_device.h>
-#include <linux/slab.h>
-
-#include <linux/firmware/qcom/qcom_scm.h>
 
 #include <soc/qcom/ice.h>
-
-#define AES_128_CBC_KEY_SIZE			16
-#define AES_256_CBC_KEY_SIZE			32
-#define AES_128_XTS_KEY_SIZE			32
-#define AES_256_XTS_KEY_SIZE			64
+#include <linux/ice_ops.h>
 
 /* QCOM ICE registers */
 #define QCOM_ICE_REG_VERSION			0x0008
@@ -45,14 +38,6 @@
 #define qcom_ice_readl(engine, reg)	\
 	readl((engine)->base + (reg))
 
-struct qcom_ice {
-	struct device *dev;
-	void __iomem *base;
-	struct device_link *link;
-
-	struct clk *core_clk;
-};
-
 static bool qcom_ice_check_supported(struct qcom_ice *ice)
 {
 	u32 regval = qcom_ice_readl(ice, QCOM_ICE_REG_VERSION);
@@ -166,178 +151,14 @@ int qcom_ice_suspend(struct qcom_ice *ice)
 }
 EXPORT_SYMBOL_GPL(qcom_ice_suspend);
 
-static int sdhci_msm_ice_set_hwkey_config(struct qcom_ice *ice,
-				          enum qcom_scm_ice_cipher cipher)
-{
-	struct device *dev = ice->dev;
-	struct ice_config_sec *ice_settings = NULL;
-	int ret;
-
-	ice_settings = kzalloc(sizeof(struct ice_config_sec), GFP_KERNEL);
-	if (!ice_settings)
-		return -ENOMEM;
-
-	switch (cipher) {
-	case QCOM_SCM_ICE_CIPHER_AES_128_XTS:
-		ice_settings->algo_mode = ICE_CRYPTO_ALGO_MODE_HW_AES_XTS;
-		ice_settings->key_size = ICE_CRYPTO_KEY_SIZE_HW_128;
-		ice_settings->key_mode = ICE_CRYPTO_USE_KEY0_HW_KEY;
-		break;
-	case QCOM_SCM_ICE_CIPHER_AES_256_XTS:
-		ice_settings->algo_mode = ICE_CRYPTO_ALGO_MODE_HW_AES_XTS;
-		ice_settings->key_size = ICE_CRYPTO_KEY_SIZE_HW_256;
-		ice_settings->key_mode = ICE_CRYPTO_USE_KEY0_HW_KEY;
-		break;
-	case QCOM_SCM_ICE_CIPHER_AES_128_ECB:
-		ice_settings->algo_mode = ICE_CRYPTO_ALGO_MODE_HW_AES_ECB;
-		ice_settings->key_size = ICE_CRYPTO_KEY_SIZE_HW_128;
-		ice_settings->key_mode = ICE_CRYPTO_USE_KEY0_HW_KEY;
-		break;
-	case QCOM_SCM_ICE_CIPHER_AES_256_ECB:
-		ice_settings->algo_mode = ICE_CRYPTO_ALGO_MODE_HW_AES_ECB;
-		ice_settings->key_size = ICE_CRYPTO_KEY_SIZE_HW_256;
-		ice_settings->key_mode = ICE_CRYPTO_USE_KEY0_HW_KEY;
-		break;
-	default:
-		dev_err_ratelimited(dev, "Unhandled cipher for HW Key support;"
-					"cipher_id=%d\n", cipher);
-		kfree(ice_settings);
-		return -EINVAL;
-	}
-	ret = qcom_config_sec_ice(ice_settings, sizeof(struct ice_config_sec));
-	kfree(ice_settings);
-
-	if (ret)
-		dev_err(dev, "Failed to program ICE HW key: %d", ret);
-
-	return ret;
-}
-
-static int qcom_ice_get_algo_mode(struct qcom_ice *ice, u8 algorithm_id,
-				  u8 key_size, enum qcom_scm_ice_cipher *cipher,
-				  u32 *key_len, bool use_hwkey)
-{
-	struct device *dev = ice->dev;
-
-	switch (key_size) {
-	case QCOM_ICE_CRYPTO_KEY_SIZE_128:
-		fallthrough;
-	case QCOM_ICE_CRYPTO_KEY_SIZE_256:
-		break;
-	default:
-		dev_err(dev, "Unhandled crypto key size %d\n", key_size);
-		return -EINVAL;
-	}
-
-	switch (algorithm_id) {
-	case QCOM_ICE_CRYPTO_ALG_AES_XTS:
-		if (key_size == QCOM_ICE_CRYPTO_KEY_SIZE_256) {
-			*cipher = QCOM_SCM_ICE_CIPHER_AES_256_XTS;
-			*key_len = AES_256_XTS_KEY_SIZE;
-		} else {
-			*cipher = QCOM_SCM_ICE_CIPHER_AES_128_XTS;
-			*key_len = AES_128_XTS_KEY_SIZE;
-		}
-		break;
-	case QCOM_ICE_CRYPTO_ALG_BITLOCKER_AES_CBC:
-		if (key_size == QCOM_ICE_CRYPTO_KEY_SIZE_256) {
-			*cipher = QCOM_SCM_ICE_CIPHER_AES_256_CBC;
-			*key_len = AES_256_CBC_KEY_SIZE;
-		} else {
-			*cipher = QCOM_SCM_ICE_CIPHER_AES_128_CBC;
-			*key_len = AES_128_CBC_KEY_SIZE;
-		}
-		break;
-	case QCOM_ICE_CRYPTO_ALG_AES_ECB:
-		/* ECB mode only supports for HW key slot */
-		if (!use_hwkey) {
-			dev_err_ratelimited(dev, "Unhandled crypto capability; "
-					"algorithm_id=%d, key_size=%d\n",
-					algorithm_id, key_size);
-			return -EINVAL;
-		}
-		if (key_size == QCOM_ICE_CRYPTO_KEY_SIZE_256) {
-			*cipher = QCOM_SCM_ICE_CIPHER_AES_256_ECB;
-			*key_len = AES_256_CBC_KEY_SIZE;
-		} else {
-			*cipher = QCOM_SCM_ICE_CIPHER_AES_128_ECB;
-			*key_len = AES_128_CBC_KEY_SIZE;
-		}
-		break;
-	default:
-		dev_err_ratelimited(dev, "Unhandled crypto capability; algorithm_id=%d, key_size=%d\n",
-				    algorithm_id, key_size);
-		return -EINVAL;
-	}
-
-	dev_info(dev, "cipher: %d key_size: %d", *cipher, *key_len);
-	return 0;
-}
-
-int qcom_ice_program_key(struct qcom_ice *ice,
-			 u8 algorithm_id, u8 key_size,
-			 const u8 crypto_key[], u8 data_unit_size,
-			 int slot, bool use_hwkey)
-{
-	struct device *dev = ice->dev;
-	enum qcom_scm_ice_cipher cipher;
-	union {
-		u8 bytes[AES_256_XTS_KEY_SIZE];
-		u32 words[AES_256_XTS_KEY_SIZE / sizeof(u32)];
-	} key;
-	int i;
-	int err;
-	u32 key_len;
-
-	if (qcom_ice_get_algo_mode(ice, algorithm_id, key_size, &cipher,
-				   &key_len, use_hwkey)) {
-		dev_err(dev, "Unhandled crypto capability; algorithm_id=%d, key_size=%d\n",
-			algorithm_id, key_size);
-		return -EINVAL;
-	}
-
-	if (use_hwkey)
-		return sdhci_msm_ice_set_hwkey_config(ice, cipher);
-
-	memcpy(key.bytes, crypto_key, key_len);
-
-	/* The SCM call requires that the key words are encoded in big endian */
-	for (i = 0; i < ARRAY_SIZE(key.words); i++)
-		__cpu_to_be32s(&key.words[i]);
-
-	err = qcom_scm_ice_set_key(slot, key.bytes, key_len, cipher,
-				   data_unit_size);
-
-	memzero_explicit(&key, sizeof(key));
-
-	return err;
-}
-EXPORT_SYMBOL_GPL(qcom_ice_program_key);
-
-int qcom_ice_evict_key(struct qcom_ice *ice, int slot)
-{
-	return qcom_scm_ice_invalidate_key(slot);
-}
-EXPORT_SYMBOL_GPL(qcom_ice_evict_key);
-
 static struct qcom_ice *qcom_ice_create(struct device *dev,
 					void __iomem *base)
 {
 	struct qcom_ice *engine;
 
-	if (!qcom_scm_is_available())
+	if (!qcom_ice_ops_available())
 		return ERR_PTR(-EPROBE_DEFER);
 
-	if (!qcom_scm_ice_available()) {
-		dev_warn(dev, "ICE SCM interface not found\n");
-		return NULL;
-	}
-
-	if (!qcom_scm_ice_hwkey_available()) {
-		dev_warn(dev, "ICE HW Key SCM interface not found\n");
-		return NULL;
-	}
-
 	engine = devm_kzalloc(dev, sizeof(*engine), GFP_KERNEL);
 	if (!engine)
 		return ERR_PTR(-ENOMEM);
diff --git a/include/linux/firmware/qcom/qcom_scm.h b/include/linux/firmware/qcom/qcom_scm.h
index 82e7ce85c64c..f1b7b1f4c3e6 100644
--- a/include/linux/firmware/qcom/qcom_scm.h
+++ b/include/linux/firmware/qcom/qcom_scm.h
@@ -401,4 +401,13 @@ extern int qti_scm_tls_hardening(uint32_t req_addr, uint32_t req_size,
 				 u32 cmd_id);
 extern int qti_scm_aes(uint32_t req_addr, uint32_t req_size, u32 cmd_id);
 extern int qti_scm_aes_clear_key_handle(uint32_t key_handle, u32 cmd_id);
+int __qcom_context_ice_sec(u32 type, u8 key_size,
+			   u8 algo_mode, u8 *data_ctxt, u32 data_ctxt_len,
+			   u8 *salt_ctxt, u32 salt_ctxt_len);
+int __qcom_config_sec_ice(void *buf, int size);
+bool __qcom_scm_ice_available(void);
+bool __qcom_scm_ice_hwkey_available(void);
+int __qcom_scm_ice_invalidate_key(u32 index);
+int __qcom_scm_ice_set_key(u32 index, const u8 *key, u32 key_size,
+			   enum qcom_scm_ice_cipher cipher, u32 data_unit_size);
 #endif
diff --git a/include/soc/qcom/ice.h b/include/soc/qcom/ice.h
index b4f58dbf74ff..83bf08f969f2 100644
--- a/include/soc/qcom/ice.h
+++ b/include/soc/qcom/ice.h
@@ -8,7 +8,12 @@
 
 #include <linux/types.h>
 
-struct qcom_ice;
+struct qcom_ice {
+	struct device *dev;
+	void __iomem *base;
+	struct device_link *link;
+	struct clk *core_clk;
+};
 
 enum qcom_ice_crypto_key_size {
 	QCOM_ICE_CRYPTO_KEY_SIZE_INVALID	= 0x0,
@@ -57,6 +62,9 @@ int qcom_ice_program_key(struct qcom_ice *ice,
 			 const u8 crypto_key[], u8 data_unit_size,
 			 int slot, bool use_hwkey);
 int qcom_ice_evict_key(struct qcom_ice *ice, int slot);
+int qcom_ice_set_context(u32 type, u8 key_size, u8 algo_mode,
+				u8 *data_ctxt, u32 data_ctxt_len,
+				u8 *salt_ctxt, u32 salt_ctxt_len);
 struct qcom_ice *of_qcom_ice_get(struct device *dev);
 struct qcom_ice *devm_of_qcom_ice_get(struct device *dev);
 
-- 
2.34.1

